etc/library.sh

```diff
@@ -114,51 +119,151 @@ function run_app_unsafe()
+      read -rp "$question" choice
+      while [[ "$choice" != "y" ]] && [[ "$choice" != "n" ]]
+      do
+        echo "choice was '$choice'"
+        read -rp "Invalid choice. y or n expected." choice
+      done
+    fi
+
+    if [[ "$choice" == "y" ]]
+    then
+      attach_to_app "$running_app"
+    fi
+    return $?
   }
 
-  # run
-  configure 2>&1 | tee -a $log
-  local ret="${PIPESTATUS[0]}"
-
+  unset configure
+  (
+    # read cfg parameters
+    [[ -f "$cfg_file" ]] && {
+      local len cfg
+      cfg="$( cat "$cfg_file" )"
+      jq -e '.tmux' <<<"$cfg" > /dev/null 2>&1
+      use_tmux="$?"
+      len="$(jq '.params | length' <<<"$cfg")"
+      for (( i = 0 ; i < len ; i++ )); do
+        local var val
+        var="$(jq -r ".params[$i].id"    <<<"$cfg")"
+        val="$(jq -r ".params[$i].value" <<<"$cfg")"
+        eval "export $var=$val"
+      done
+    }
+ 
+    echo "$ncp_app" > "$LOCK_FILE"
+    if which tmux > /dev/null && [[ $use_tmux == 0 ]]
+    then
+      echo "Running $ncp_app in tmux..." | tee -a $log
+      # Run app in tmux
+      local tmux_log_file tmux_status_file LIBPATH
+      tmux_log_file="/var/log/ncp/tmux.${ncp_app}.log"
+      tmux_status_file="/var/log/ncp/tmux.${ncp_app}.status"
+      LIBPATH="$(dirname $CFGDIR)/library.sh"
+      
+      # Reset tmux output
+      echo "[ $ncp_app ]" | tee -a $log
+      echo "[ $ncp_app ]" > "$tmux_log_file"
+      echo "" > "$tmux_status_file"
+      chmod 640           "$tmux_log_file" "$tmux_status_file"
+      chown root:www-data "$tmux_log_file" "$tmux_status_file"
+
+      tmux new-session -d -s "$ncp_app" "bash -c '(
+        trap \"echo \\\$? > $tmux_status_file && rm $LOCK_FILE\" 1 2 3 4 6 9 11 15 19 29
+        source \"$LIBPATH\"
+        source \"$script\"
+        configure 2>&1 | tee -a $log
+        echo \"\${PIPESTATUS[0]}\" > $tmux_status_file
+        rm $LOCK_FILE
+      )' 2>&1 | tee -a $tmux_log_file"
+
+      attach_to_app "$ncp_app"
+      exit
+
+    else
+      trap "rm '$LOCK_FILE'" 0 1 2 3 4 6 11 15 19 29
+      echo "[ $ncp_app ]" | tee -a $log
+      echo "Running $ncp_app directly..." | tee -a $log
+      # read script
+      # shellcheck source=/dev/null
+      source "$script"
+      # run
+      configure 2>&1 | tee -a $log
+      local ret="${PIPESTATUS[0]}"
+      exit "$ret"
+    fi
+  )
+  ret="$?"
   echo "" >> $log
 
   return "$ret"
 }
 
+function attach_to_app()
+{
+  local tmux_log_file tmux_status_file
+  tmux_log_file="/var/log/ncp/tmux.${ncp_app}.log"
+  tmux_status_file="/var/log/ncp/tmux.${ncp_app}.status"
+
+  if [[ "$ATTACH_NO_FOLLOW" == "1" ]]
+  then
+    cat "$tmux_log_file"
+    return 0
+  else
+    (while tmux has-session -t="$ncp_app" > /dev/null 2>&1 
+    do
+      sleep 1
+    done) &
+
+    # Follow log file until tmux session has terminated
+    tail --lines=+0 -f "$tmux_log_file" --pid="$!"
+  fi
+
+  # Read return value from tmux log file
+  ret="$(tail -n 1 "$tmux_status_file")"
+  #rm "$tmux_log_file"
```

Comment:
comments

Reply:
Fixed in 2a4ee286d0ef58a8cbfe5a840f89a6714f48f41a